### PR TITLE
Adding recipe list and create endpoints

### PR DESCRIPTION
**Pull request by omerfarukabaci**

### CODE REVIEW

File: recipes/serializers.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        recipe.save()
+        for ingredient_name in data["ingredients"]:
+            lookup_name = ingredient_name.lower()
+            lookup_name = lookup_name.replace(" ", "_")
+            try:
+                ingredient = Ingredient.objects.get(lookup_name=lookup_name)
+                recipe.ingredients.add(ingredient)
+            except ObjectDoesNotExist:
+                ingredient = Ingredient(name=ingredient_name, lookup_name=lookup_name)
+                ingredient.save()
```
Comment (alicertel) [inline recipes/serializers.py]: You can use get_or_create

File: recipes/serializers.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                recipe.ingredients.add(ingredient)
+
+        recipe.difficulty = str(data["difficulty"])
+        recipe.save()
```
Comment (alicertel) [inline recipes/serializers.py]: why do you save recipe twice?

Comment (omerfarukabaci) [reply recipes/serializers.py]: Since Recipe have a many to many relationship with Ingredient, it has to have an id before adding ingredient instances to it. That's why I have used .save() twice.

File: recipes/serializers.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        return recipe
+
+    def to_representation(self, data):
```
Comment (alicertel) [inline recipes/serializers.py]: why do you override to_representation?

File: recipes/serializers.py (hunk 2 above)
Comment (alicertel) [reply recipes/serializers.py]: This can be achieved without saving recipe twice.

File: recipes/urls_api.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+from . import views_api
+
+app_name = 'api-recipes'
+
+urlpatterns = [
+    path('recipes/', views_api.RecipesView.as_view(), name="recipes"),
```
Comment (alicertel) [inline recipes/urls_api.py:7]: Nice.

File: recipes/serializers.py (hunk 3 above)
Comment (omerfarukabaci) [reply recipes/serializers.py]: Here data (which is a Recipe object) have a many to many field which points to a list of Ingredient, and the list of Ingredient objects is not JSON serializable so I had to override to_representation to get and print ingredient names.

Comment (alicertel) [reply recipes/serializers.py]: I believe you can find a better solution to this without overriding the to_representation method.

File: hipo_meal/settings.py [hunk 5]
```diff
@@ -144,5 +144,6 @@
     ),
     'DEFAULT_PERMISSION_CLASSES': (
         'rest_framework.permissions.IsAuthenticatedOrReadOnly',
-    )
+    ),
+    'DATETIME_FORMAT': "%Y-%m-%d %H:%M",
```
Comment (alicertel) [inline hipo_meal/settings.py]: Formating the date is work of the clients. By default `iso-8601` is used by drf. It returns timezone aware date time object. This allow clients to format it in any way they want.

File: recipes/serializers.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        model = Ingredient
+        fields = ('name',)
+        extra_kwargs = {
+            'name': {'validators': []},
+        }
+
+
+class RecipeSerializer(serializers.ModelSerializer):
+    ingredients = IngredientSerializer(many=True)
+    difficulty = ChoiceField(choices=Recipe.DIFFICULTY_CHOICES)
```
Comment (alicertel) [inline recipes/serializers.py:21]: You don't need a custom field for this. You can use serializers.CharField(choices=...)

File: recipes/serializers.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+class ChoiceField(serializers.ChoiceField):
+    def to_representation(self, obj):
+        return self._choices[obj]
+
+
+class IngredientSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Ingredient
+        fields = ('name',)
+        extra_kwargs = {
```
Comment (alicertel) [inline recipes/serializers.py:14]: why did you remove the validator of name field?

File: recipes/serializers.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            'name': {'validators': []},
+        }
+
+
+class RecipeSerializer(serializers.ModelSerializer):
+    ingredients = IngredientSerializer(many=True)
+    difficulty = ChoiceField(choices=Recipe.DIFFICULTY_CHOICES)
+    author = serializers.SerializerMethodField(source='get_author')
+    date_posted = serializers.DateTimeField(required=False)
```
Comment (alicertel) [inline recipes/serializers.py]: this should be read only field.

File: recipes/serializers.py (hunk 7 above)
Comment (omerfarukabaci) [reply recipes/serializers.py:14]: If I don't it checks whether the ingredient already exists or not and if so it raises a validation error, which prevent me adding new recipes with existing ingredients. I don't want this validation check since in create method I manually check if an ingredient exists or not. Some validators still works though (e.g. allow_blank).

**Comment by alicertel** [comment]
Thank you. Looks great. Ready to merge.
